incubator/hnc/internal/pkg/selectors/selectors.go

```diff
@@ -18,33 +19,62 @@ func GetSelectorAnnotation(inst *unstructured.Unstructured) string {
 	return annot[api.AnnotationSelector]
 }
 
+func GetTreeSelectorAnnotation(inst *unstructured.Unstructured) string {
+	annot := inst.GetAnnotations()
+	return annot[api.AnnotationTreeSelector]
+}
+
 // GetTreeSelector is similar to a regular selector, except that it adds the LabelTreeDepthSuffix to every string
 // To transform a tree selector into a regular label selector, we follow these steps:
 // 1. get the treeSelector annotation if it exists
 // 2. convert the annotation string to a slice of strings seperated by comma, because user is allowed to put multiple selectors
 // 3. append the LabelTreeDepthSuffix to each of the treeSelector string
 // 4. combine them into a single string connected by comma
 func GetTreeSelector(inst *unstructured.Unstructured) (labels.Selector, error) {
-	annot := inst.GetAnnotations()
-	treeSelectorStr, ok := annot[api.AnnotationTreeSelector]
-	if !ok {
+	treeSelectorStr := GetTreeSelectorAnnotation(inst)
+	if treeSelectorStr == "" {
 		return nil, nil
 	}
-	strs := strings.Split(treeSelectorStr, ",")
+
+	segs := strings.Split(treeSelectorStr, ",")
 	selectorStr := ""
-	for i, str := range strs {
-		selectorStr = selectorStr + str + api.LabelTreeDepthSuffix
-		if i < len(strs)-1 {
+	for i, seg := range segs {
+		// check if it's a valid namespace name
+		if err := validateNamespaceName(seg); err != nil {
+			return nil, err
+		}
+
+		selectorStr = selectorStr + seg + api.LabelTreeDepthSuffix
+		if i < len(segs)-1 {
 			selectorStr = selectorStr + ", "
 		}
 	}
 	treeSelector, err := getSelectorFromString(selectorStr)
 	if err != nil {
-		return nil, fmt.Errorf("while parsing %q: %w", api.AnnotationTreeSelector, err)
+		// In theory this should never happen because we already checked DNS label before.
+		// If this happens, it's more likely that we have a bug in our code
+		return nil, fmt.Errorf("Internal error while parsing %q: %w", api.AnnotationTreeSelector, err)
```

Comment:
Error usually start with a lowercase letter since they'll often get wrapped, like you did here :)

Log message, by contrast, start with upper case letters because they can't be wrapped (same with user messages).

Reply:
Good to know! Corrected